File: product-service/src/main/java/ru/otus/spring/belov/product_service/service/CategoryService.java
package ru.otus.spring.belov.product_service.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import ru.otus.spring.belov.product_service.dto.category.CategoryFilter;
import ru.otus.spring.belov.product_service.dto.category.CategoryTreeItem;
import ru.otus.spring.belov.product_service.dto.category.CategoryItem;
import ru.otus.spring.belov.product_service.dto.category.SaveCategoryRequest;

import java.util.List;

/**
 * Сервис по работе с категориями
 */
public interface CategoryService {

    /**
     * Возвращает дерево категорий
     * @param categoryFilter фильтр
     * @return дерево категорий
     */
    List<CategoryTreeItem> getCategoriesTree(CategoryFilter categoryFilter);

    /**
     * Возвращает категории из корзины
     * @return категории из корзины
     */
    Page<CategoryItem> getTrash(Pageable pageable);

    /**
     * Переводит категории в корзину
     * @param ids идентификаторы категорий
     */
    void moveToTrash(List<Long> ids);

    /**
     * Переводит категории из корзину
     * @param ids идентификаторы категорий
     */
    void restoreTrash(List<Long> ids);

    /**
     * Удаляет категории
     * @param ids идентификаторы категорий
     */
    void delete(List<Long> ids);

    /**
     * Возвращает сохранённую/изменённую категорию
     * @param saveCategoryRequest запрос на изменение/сохранение
     */
    CategoryItem saveCategory(SaveCategoryRequest saveCategoryRequest);

    /**
     * Возвращает категорию
     * @param catId идентификатор категории
     * @return категория
     */
    CategoryItem getCategoryById(Long catId);
}

File: admin-ui/src/ts/services/CategoryService.ts
import {OnlyInstantiableByContainer, Singleton} from "typescript-ioc";
import axios from "axios";
import {SpringPageable, TableData} from "@/ts/components/common/PageableTable";

/**
 * Сервис по работе с авторизацией
 */
@Singleton
@OnlyInstantiableByContainer
export class CategoryService {

    /**
     * Возвращает категории в виде дерева
     */
    async getCategoriesTree(): Promise<CategoryTreeItem[]> {
        return (await axios.get<CategoryTreeItem[]>("/product-service/admin/categories/tree", {})).data;
    }

    /**
     * Возвращает категорию
     * @param id идентификатор категории
     */
    async getCategory(id: number): Promise<CategoryItem> {
        return (await axios.get<CategoryItem>(`/product-service/categories/${id}`)).data;
    }

    /**
     * Возвращает содержимое корзины
     */
    async getTrash(pageable: SpringPageable): Promise<TableData<CategoryItem>> {
        return (await axios.get<TableData<CategoryItem>>("/product-service/admin/categories/trash", {
            params: pageable
        })).data;
    }

    /**
     * Переносит в корзину
     * @param ids идентификаторы категории
     */
    async moveToTrash(ids: number[]): Promise<void> {
        await axios.post("/product-service/admin/categories/trash", ids);
    }

    /**
     * Переносит в корзину
     * @param ids идентификаторы категории
     */
    async trashRestore(ids: number[]): Promise<void> {
        await axios.post("/product-service/admin/categories/trash/restore", ids);
    }

    /**
     * Удаляет с концами
     * @param ids идентификаторы категории
     */
    async delete(ids: number[]): Promise<void> {
        await axios.delete("/product-service/admin/categories/", {
            data: ids
        });
    }

    /**
     * Сохраняет категорию
     * @param categoryItem сохраняемая категория
     */
    async saveCategory(categoryItem: CategoryItem): Promise<CategoryItem> {
        return (await axios.post<CategoryItem>("/product-service/admin/categories/", categoryItem)).data;
    }
}

/** Список категорий в виде дерева */
export type CategoryTreeItem = {

    /** Идентификатор */
    id: number;
    /** Заголовок */
    title: string;
    /** Признак что ресурс в корзине */
    deleted: boolean;
    /** Скрыт в меню */
    hideMenu: boolean;
    /** Опубликован */
    published: boolean;
    /** Индекс сортировки */
    sortIndex: number;
    /** Дочерние категории */
    childs: CategoryTreeItem[];
}

/** Категория на редактирование */
export type CategoryItem = {
    /** Идентификатор */
    id: number | null;
    /** Заголовок */
    title: string;
    /** Признак что ресурс в корзине */
    deleted: boolean;
    /** Скрыт в меню */
    hideMenu: boolean;
    /** Опубликован */
    published: boolean;
    /** Индекс сортировки */
    sortIndex: number;
    /** Родительская категория */
    parent: number | null;
}